### DISCUSSION TITLE
Exploiting Task Parallelism with OpenCL

### DISCUSSION THREAD

**Discussion by mm-harish** [pocl/pocl#1707] [Q&A] [answered]
I'm currently reading the paper Exploiting Task Parallelism with OpenCL: A Case Study. Is it been implemented on PoCL and if so, can someone point me to the code. Specifically on where the Graph is being constructed.

**Comment by pjaaskel** [comment, marked as answer]
The code has changed and evolved (also features removed) quite a bit since that publication. In that publication's time we didn't have command buffers and PoCL attempted to optimize at "graph level" despite this. We are now moving towards utilizing the multi-device command buffer extension for optimizing heterogeneous task graphs. Anyhow, I think implicit task parallelism extracted from in-order-queues' buffer dependencies is still there as we have quite sophisticated (sub)buffer analysis in place.
